src/main/webapp/script.js

```diff
@@ -149,39 +159,229 @@ class AutocompleteDirectionsHandler {
+                // Or else no data available and will have error saying it is undefined (because it doesnt exist). 
+                // If the mapRow from getGridIndex of stepAQIdoesnt doesn't exist in the data grid, stepAQI is 0 for now. 
+                let stepAQI = 0;
+                if (!dataGrid[mapRow]) {
+                    console.log("Row map doesnt exist and is undefined.");
+                } else if (!dataGrid[mapRow][mapCol]) {
+                    console.log("Step aqi dosnt exist because mapCol doesn't exist in mapRow. So cant find and set stepAQI");
+                } else {
+                    stepAQI = dataGrid[mapRow][mapCol];
+                }
+
+                console.log(stepAQI);
+                totalValue += stepWeight * stepAQI;
+                totalWeight += stepWeight;
+                console.log("So far total value is: " + totalValue + " total weight is: "+ totalWeight);
+            }
+
+            // Counting the AQI at the end point of the leg as part of the score. 
+            // Set duration stepWeight as 60 seconds to account of parking upon arrival at end point. 
+            // TODO: Rosanna to rethink on this logic, would it stll apply with waypoints where there are a lot of legs? It can count for stopover time. 
+            let legWeight = 60;
+            let legEndlat = leg["end_location"].lat();
+            let legEndLng = leg["end_location"].lng();
+
+            let endPtRowCol = getGridIndex(legEndlat, legEndLng, aqDataPointsPerDegree);
+            let mapRow = endPtRowCol.row;
+            let mapCol = endPtRowCol.col;
+            let endptAQI = 0;
+
+            // Print to see in console whether mapRow or mapCol is undefined. 
+            if (!dataGrid[mapRow]) {
+                console.log("Row map doesnt exist and is undefined.");
+            } else if (!dataGrid[mapRow][mapCol]) {
+                console.log("Step aqi dosnt exist because mapCol doesn't exist in mapRow. So cant find and set stepAQI");
+            } else {
+                endptAQI = dataGrid[mapRow][mapCol];
+            }
+
+            console.log(endptAQI);
+            totalValue += legWeight * endptAQI;
+            totalWeight += legWeight;
+            console.log("So far total value is: " + totalValue + " total weight is: "+ totalWeight);
+        }
+
+        // Calculate total route Score 
+        let routeScore = totalValue / totalWeight; 
+        console.log("Total value is: " + totalValue + " totalweight is: " + totalWeight);
+        console.log("The Route Score is: " + routeScore);
+        return routeScore;
+    }
+
     calculateAndDisplayRoute() {
+        // If PlaceIDs are not placed in yet, dont calculate a route yet. 
         if (!this.originPlaceId || !this.destinationPlaceId) {
             return;
         }
+        const waypts = [];
+        const wayptAutocompleteArray = this.waypointAutocompleteArray;
+
+        for (const autocompleteObject of wayptAutocompleteArray) {
```

Comment:
why do you need this variable, shall we use `this. waypointAutocompleteArray` instead?

Reply:
i had this variable in cose I thought it might bring better readability but in hindsight there isnt a huge different aye :")